enhancements/baremetal/ztp-metal3.md

```diff
@@ -0,0 +1,239 @@
+  `BareMetalHost`s it manages, will start using a different process based
+  on the  new `install_assisted` deploy method and stop rebooting the machine
+  in the end of the process.
+
+### User Stories
+
+N/A (this change itself does not add new user stories)
+
+### API Extensions
+
+All required API extensions are already present.
+
+### Implementation Details/Notes/Constraints
+
+#### Image building
+
+ICC will follow different code paths depending on whether the `BareMetalHost`
+has the `infraenvs.agent-install.openshift.io` annotation linking it to an
+`InfraEnv` resource. In both cases it will generate an Ignition configuration
+to start and configure the Ironic agent. In the ZTP case, the corresponding
+`InfraEnv` will be updated with the generated Ignition configuration (merging
+it with any user-provided configuration).
+
+The same process will be applied to `initrd` images for (i)PXE booting. The
+Ignition configuration will be embedded into the resulting `initrd` by the
+assisted service. The root filesystem (required for CoreOS) will be provided
+by [coreos-downloader][coreos-downloader], similarly to the *standard* flow.
+
+#### Coordination between agents
+
+The first agent to start will be the Ironic agent since it is responsible
+for coordinating the whole flow and (indirectly through Ironic) updating the
+`BareMetalHost`'s status. *Inspection* and additional configuration (RAID etc)
+will be handled by the Ironic agent alone.
+
+Currently, the installation is initiated by applying the following change to
+the `BareMetalHost`'s spec:
+
+```yaml
+Image:
+  URL: url/of/assisted/agent.iso
+  DiskFormat: live-iso
+```
+
+The new process will be instead triggered by:
+
+```yaml
+CustomDeploy:
+  Method: install_assisted
+```
+
+This instructs Ironic to find and execute a *custom deploy step* called
+`install_assisted`, which will be provided by our Ironic agent plugin. This
+step will use D-Bus to order systemd to start the assisted agent and watch
+its status.
+
+If the assisted agent exits successfully, the Ironic agent will report success
+to Ironic and baremetal-operator. Otherwise, a failure will be recorded.
+
+### Risks and Mitigations
+
+- More containers will be downloaded on the ramdisk start-up. We'll need to
+  account for more RAM and a longer start-up time.
+- A user with editing access to `InfraEnv` resources can potentially break
+  the Ironic agent configuration. Short-term, we'll probably document this
+  as a known issue. Long-term, we may need to update `InfraEnv` with a new
+  field.
+
+## Design Details
+
+### Open Questions
+
+- We are going to put the user's pull secret directly to the `InfraEnv`
+  resources as part of the Ignition override. Is it acceptable?
+- Is there a chance of version mismatch between the CoreOS `initrd` from the
+  assisted service and the root filesystem provided by
+  [coreos-downloader][coreos-downloader]?
+
+### Test Plan
+
+The feature will be covered by the existing ZTP e2e tests.
```

Comment:
It may be helpful to identify and list the ZTP e2e tests which currently exist - do we need some new ones?

For example should we have CI coverage for the newly-enabled PXE flow and/or RAID/firmware config?

How can we validate on upgrade that existing ZTP deployments aren't impacted (maybe this can be manual QE validation vs CI testing)?

Reply:
I'll write a few basic things here, but I'd like input from assisted folks.